feat(viewer): add Text, Markdown and Source Code viewer (read-only)

ShGKme opened the pull request:
- Based on https://github.com/nextcloud/talk-desktop/pull/784
- Only the last commit is related

### ☑️ Resolves

* A part of: https://github.com/nextcloud/talk-desktop/issues/256

Includes features:
- File types:
  - `.md` with extended markdown support
  - `.txt` with auto link only
  - Different source code files with monospaced font (depends on the server's mimetype detection)
- Switch between 900px / fullwidth modes
- Copying text content
- Navigating/selecting like in a readonly contenteditable area
- Handling empty files

### 🖼️ Screenshots

#### Text file

🌕 Light | 🌑 Dark
---|---
![image](https://github.com/user-attachments/assets/4086536e-6687-4fbc-ad4d-8915f9413911) | ![image](https://github.com/user-attachments/assets/4eb704fd-5245-4570-8dd5-8684cf1a7843)

#### Markdown file

🌕 Light | 🌑 Dark
---|---
![image](https://github.com/user-attachments/assets/9e3ae4dc-41f9-4bdb-8c3e-f89976a23cf4) | ![image](https://github.com/user-attachments/assets/3c92f4a3-f38e-47a6-bbbb-ec41ad9fa419)

#### Full-width mode

![image](https://github.com/user-attachments/assets/fdcfbcbb-1f85-4335-a4e8-479ea71b31d7)

#### Soft-wrap - MD

Enabled | Disabled
---|---
![image](https://github.com/user-attachments/assets/45e21d27-bd88-47bf-a08d-6df9ebda9223) | ![image](https://github.com/user-attachments/assets/5da142ee-aa35-4bf5-ba6b-9219cdcfac53)
![image](https://github.com/user-attachments/assets/9b17a4e0-b569-4e00-bb5c-ab392b23f968) | ![image](https://github.com/user-attachments/assets/c1daa65f-54a4-4f28-bff2-6ee9c59bec91)

#### Code (no highlighting)

🌕 Light | 🌑 Dark
---|---
![image](https://github.com/user-attachments/assets/09f03c20-773c-47fd-ad75-baac0c37e997) | ![image](https://github.com/user-attachments/assets/4c251f31-9fb4-4c77-a1e9-5f82bd1e59e2)

#### Empty file

🌕 Light | 🌑 Dark
---|---
![image](https://github.com/user-attachments/assets/926509df-03ec-4b6d-a15e-52c8656b592a) | ![image](https://github.com/user-attachments/assets/50a39bac-a069-4188-aa2d-5c54d37e2a84)

#### Copy content

![copy](https://github.com/user-attachments/assets/66ab0add-90f1-4960-8d34-c4e0d779d6f4)

#### Navigation

The text is within `contenteditable`, so a user can select/copy/navigate like in a text editor. But it is readonly.

![navigation](https://github.com/user-attachments/assets/3a5381e5-81ab-4978-9041-00fbbc06b2ba)

Antreesy commented on src/talk/renderer/Viewer/ViewerHandlerText.vue:
```diff
@@ -0,0 +1,182 @@
+import NcCheckboxRadioSwitch from '@nextcloud/vue/dist/Components/NcCheckboxRadioSwitch.js'
+import NcEmptyContent from '@nextcloud/vue/dist/Components/NcEmptyContent.js'
+import NcRichText from '@nextcloud/vue/dist/Components/NcRichText.js'
+import IconCropPortrait from 'vue-material-design-icons/NoteTextOutline.vue'
```
Maybe this icon would fit better (rotated/mirrored)? NoteText looks too squarish

```css
    transform: rotate(90deg) scaleX(-1);
```

<img width="69" alt="image" src="https://github.com/user-attachments/assets/f91217ec-016a-401b-b7f9-3a9e64bd8692">

https://pictogrammers.com/library/mdi/icon/file-chart-outline/

ShGKme replied:
Smart

Antreesy commented on src/talk/renderer/Viewer/ViewerHandlerText.vue:
```diff
@@ -0,0 +1,182 @@
+							`viewer-text__content--${format}`, {
+								'viewer-text__content--compact': layout === 'compact',
+							}]"
+						contenteditable
+						spellcheck="false"
+						@beforeinput.prevent>
```
Is that needed, if we have read-only?

ShGKme replied:
We don't have readonly and it is not supported on `[contenteditable]`. So we have to emulate readonly...

ShGKme commented:
Added many fixes separated into small commits. See PR description for new images.

- More types for code from the suggestion
- Changed icon from the suggestion
- Fix Copy button (it wasn't actually implemented, lol)
- Updated design to handle light theme
- Improved wrap/non-wrap text view + added a control for it
- Made menubar sticky on top instead of fixed on the corner
- Some minor UI changes

ShGKme commented on src/talk/renderer/Viewer/ViewerHandlerText.vue:
```diff
@@ -0,0 +1,182 @@
+
+.viewer-text__content--txt,
+.viewer-text__content--code,
+.viewer-text__content--md :deep(pre) {
```
Fixed

DorraJaouad commented on src/talk/renderer/Viewer/ViewerHandlerText.vue:
```diff
@@ -0,0 +1,210 @@
+	},
+})
+
+const format = computed(() => props.file.mime === 'text/markdown' ? 'md' : 'txt')
```
You missed `code` returned value

Antreesy replied:
Now overflow content for code is scrollable, but not visible 🙈 

![2024-09-05_17h24_05](https://github.com/user-attachments/assets/5ab59414-4ec6-4a2b-a52c-d7f2954e57ac)

ShGKme replied:
(ヘ･_･)ヘ┳━┳

(╯°□°）╯︵ ┻━┻

ShGKme replied:
Fixed

DorraJaouad commented on src/talk/renderer/Viewer/ViewerHandlerText.vue:
```diff
@@ -0,0 +1,210 @@
+
+.viewer-text__controls {
+	background-color: inherit;
+	width: 900px;
```
Maybe better with full width?

ShGKme replied:
Small | Large
-- | --
![image](https://github.com/user-attachments/assets/918f745d-997c-42b8-9878-1704728699f8) | ![image](https://github.com/user-attachments/assets/fbc7da32-fbad-4ce4-8dbc-11d162e977ae)

Antreesy replied:
Doesn't look good at light theme, with BG color filling the whole area:

<img width="719" alt="image" src="https://github.com/user-attachments/assets/e266e9ce-03ef-4dc0-911b-6af1deb01f68">

Here I applied `max-width: 900px`:
<img width="648" alt="image" src="https://github.com/user-attachments/assets/a8068f47-fe12-44b9-add4-871c88ba8a09">

DorraJaouad replied:
Because when you have full width mode, and you have text background different (e.g: code) the side edges of 900px bar will be seen.

Antreesy replied:
8px side paddings seems to be fine.
<img width="704" alt="image" src="https://github.com/user-attachments/assets/81973bd5-3ba7-4f02-9ec7-8c0b6a22873c">

Not sure how it would look like with `format === code`, because it's not working anymore 👀, but IMO substrate layer with BG color should be applied to all kinds of files (respecting compact mode), and then `--dark-background` should go on top of it

ShGKme replied:
background was moved from text content only to an entire container to be consistent with the Text app

ShGKme replied:
![image](https://github.com/user-attachments/assets/0eeb1a33-1f44-4157-b32e-2a1547c98f41)

ShGKme replied:
![image](https://github.com/user-attachments/assets/5ef1336f-20ea-4513-a688-3a9739d1d557)

Antreesy replied:
> to be consistent with the Text app

Ah, okay)

ShGKme commented on src/talk/renderer/Viewer/ViewerHandlerText.vue:
Hmm, a text file has no ETag change after its content was changed... So technically the cached value is valid...

ShGKme replied:
Checking with the office team

ShGKme replied:
Works fine if reopen after some short time, just not immediately after is was changes in the web-editor.

ShGKme commented:
- Fixed regression with no `code` format
- Moved button to the right (removed 900px limitation)

Antreesy reviewed (approved):
Good from my POV
